static/app/components/searchQueryBuilder/selectionKeyHandler.tsx

```diff
@@ -0,0 +1,182 @@
+import {type ForwardedRef, forwardRef, useCallback} from 'react';
+import {VisuallyHidden} from '@react-aria/visually-hidden';
+import type {ListState} from '@react-stately/list';
+import type {Key} from '@react-types/shared';
+
+import {useSearchQueryBuilder} from 'sentry/components/searchQueryBuilder/context';
+import {type ParseResultToken, Token} from 'sentry/components/searchSyntax/parser';
+import {defined} from 'sentry/utils';
+import {isCtrlKeyPressed} from 'sentry/utils/isCtrlKeyPressed';
+
+type SelectionKeyHandlerProps = {
+  state: ListState<ParseResultToken>;
+  undo: () => void;
+};
+
+function findNearestFreeTextKey(
+  state: ListState<ParseResultToken>,
+  startKey: Key | null,
+  direction: 'right' | 'left'
+): Key | null {
+  let key: Key | null = startKey;
+  while (key) {
+    const item = state.collection.getItem(key);
+    if (!item) {
+      break;
+    }
+    if (item.value?.type === Token.FREE_TEXT) {
+      return key;
+    }
+    key = (direction === 'right' ? item.nextKey : item.prevKey) ?? null;
+  }
+
+  if (key) {
+    return key;
+  }
+
+  return direction === 'right'
+    ? state.collection.getLastKey()
+    : state.collection.getFirstKey();
+}
+
+/**
+ * SelectionKeyHandler is used to handle keyboard events when a selection is
+ * active, which differ from default behavior. When the user has a selection,
+ * this component should be focused.
+ *
+ * We use an invisible input element in order to handle paste events. Without
+ * this, the browser will need to ask for clipboard permissions.
+ */
+export const SelectionKeyHandler = forwardRef(
+  ({state, undo}: SelectionKeyHandlerProps, ref: ForwardedRef<HTMLInputElement>) => {
+    const {dispatch} = useSearchQueryBuilder();
+
+    const selectedTokens = Array.from(state.selectionManager.selectedKeys)
+      .map(key => state.collection.getItem(key)?.value)
+      .filter(defined);
+
+    const onPaste = useCallback(
+      (e: React.ClipboardEvent<HTMLInputElement>) => {
+        e.preventDefault();
+        e.stopPropagation();
+
+        const text = e.clipboardData.getData('text/plain').replace('\n', '').trim();
+
+        dispatch({
+          type: 'REPLACE_TOKENS_WITH_TEXT',
+          tokens: selectedTokens,
+          text,
+        });
+      },
+      [dispatch, selectedTokens]
+    );
+
+    const onKeyDown = useCallback(
```

Comment:
All this logic is copied over from `useQueryBuilderGrid`, the only new part of this component is the rendered input and the `onPaste` handler